python/cudf/cudf/_lib/pylibcudf/column.pyx

```diff
@@ -0,0 +1,95 @@
+# Copyright (c) 2023, NVIDIA CORPORATION.
+
+from cython.operator cimport dereference
+from libcpp.memory cimport unique_ptr
+from libcpp.utility cimport move
+from libcpp.vector cimport vector
+
+from rmm._lib.device_buffer cimport DeviceBuffer
+
+from cudf._lib.cpp.column.column cimport column, column_contents
+from cudf._lib.cpp.types cimport bitmask_type, offset_type, size_type
+
+from .gpumemoryview cimport gpumemoryview
+from .types cimport DataType
+from .utils cimport int_to_bitmask_ptr, int_to_void_ptr
+
+
+cdef class Column:
+    """A container of nullable device data as a column of elements."""
+    def __init__(
+        self, DataType data_type not None, size_type size, gpumemoryview data,
+        gpumemoryview mask, size_type null_count, offset_type offset,
+        # TODO: Not sure what the best input is for children, for now just
+        # using a List[Column]
+        object children
+    ):
+        self.data_type = data_type
+        self.size = size
+        self.data = data
+        self.mask = mask
+        self.null_count = null_count
+        self.offset = offset
+        self.children = children
+
+    cdef column_view* get_underlying(self):
+        cdef const void * data = NULL
+        cdef const bitmask_type * null_mask = NULL
+        cdef vector[column_view] c_children
+        cdef Column child
+
+        if not self._underlying:
+            if self.data is not None:
+                data = int_to_void_ptr(self.data.ptr)
+            if self.mask is not None:
+                null_mask = int_to_bitmask_ptr(self.mask.ptr)
+
+            if self.children is not None:
+                for child in self.children:
+                    c_children.push_back(dereference(child.get_underlying()))
+
+            self._underlying.reset(
+                new column_view(
+                    self.data_type.c_obj, self.size, data, null_mask,
+                    self.null_count, self.offset, c_children
+                )
+            )
+        return self._underlying.get()
```

Comment:
Is there a reason you store a `unique_ptr[column_view]` and then hand out the raw pointer, rather than the C++ version which just hands out a new `column_view` (no pointers anywhere) each time?

Reply:
Yeah, there are a couple of reasons (not necessarily great ones):
1. The way that Cython transpiles to C++ usually adds a bunch of extra intermediate assignments. In C++ if we have a method `cudf::foo(column_view x)` and do something like `column x = ...; cudf::foo(x->view())`, the compiler would typically elide the copy i.e. the temporary created by `x->view()` wouldn't be copied _again_ when passed as an argument to `cudf::foo` because the compiler would see that the temporary was being immediately consumed so it would reuse the object. However, with Cython it's likely that something like `cudf.foo(x.view())` will actually translate to something like `column_view y; ...; y = x->view(); ...; cudf::foo(y)`, and as a result copy elision will probably not occur. I chose the current approach because it guarantees not copying column views all over the place.
2. There are restrictions on what types of code can be run in `nogil` blocks. The `get_underlying` method needs to access the list (a Python object) of children in order to construct the `cudf::column_view`. That means that this function cannot be marked nogil (not quite true, you could mark it `nogil` and then use a `with gil` block inside it, but I didn't want to do anything that fancy yet without good reason). That in turn means that we can't call `get_underlying` directly inside the `with nogil` blocks that we typically use in the algorithms calling to libcudf code. That pretty much guarantees that we can't get copy elision working here, so using a pointer is pretty much the only way to guarantee reducing copies.

I say these aren't great reasons because this certainly gets back to the point about whether we are doing things to work around Cython limitations. That said, I'm not sure pybind11/nanobind would necessary fix this altogether, you might just trade them in for different problems (I haven't worked pybind in a couple years now, and I've never used nanobind, only seen its code).

Reply:
I've added documentation of this behavior so that it's clear enough in case we want to reverse this decision later.

Reply:
OK, I've actually decided to walk this back. Adding a little extra complexity to how the conversion works was enough to make it possible to return a new `column_view` by value each time this function is created, and to ensure that we could write Cython code in a way that favors copy elision.